webapp/src/lib/erc20.ts

```diff
@@ -0,0 +1,125 @@
+import { Contract, type Provider as EthersProvider } from "ethers"; // Adjust Provider type as needed
+
+export interface ERC20TokenDetails {
+  address: string;
+  name: string;
+  symbol: string;
+  decimals: number;
+  balance: bigint;
+}
+
+export const ERC20_ABI = [
+  "function name() view returns (string)",
+  "function symbol() view returns (string)",
+  "function decimals() view returns (uint8)",
+  "function balanceOf(address account) view returns (uint256)",
+];
+
+/**
+ * Fetches details (name, symbol, decimals) and balance for an ERC20 token.
+ * @param provider Ethers.js provider instance.
+ * @param tokenAddress The address of the ERC20 token contract.
+ * @param ownerAddress The address of the owner for whom to fetch the balance.
+ * @returns A promise that resolves to an ERC20TokenDetails object or null if an error occurs.
+ */
+export async function fetchERC20TokenDetails(
+  provider: EthersProvider, // Using the imported Provider type alias
+  tokenAddress: string,
+  ownerAddress: string,
+): Promise<ERC20TokenDetails | null> {
+  try {
+    // Basic validation for addresses
+    if (!tokenAddress.match(/^0x[a-fA-F0-9]{40}$/)) {
+      console.error(`Invalid token address format: ${tokenAddress}`);
+      return null;
+    }
+    if (!ownerAddress.match(/^0x[a-fA-F0-9]{40}$/)) {
+      console.error(`Invalid owner address format: ${ownerAddress}`);
+      return null;
+    }
+
+    const tokenContract = new Contract(tokenAddress, ERC20_ABI, provider);
+
+    // Validate if the address is a contract
+    const code = await provider.getCode(tokenAddress);
+    if (code === '0x' || code === null) { // code can also be null if network error or other issues
+      console.error(`Address ${tokenAddress} is not a contract or unable to fetch code.`);
+      return null;
+    }
+
+    // Using Promise.allSettled to get all results even if one fails, though individual call errors are caught by the outer try/catch here.
+    // For this specific case, Promise.all is fine because if one essential detail (like name or symbol) fails, 
+    // the whole token detail object is likely invalid.
+    const [nameResult, symbolResult, decimalsResult, balanceResult] = await Promise.allSettled([
```

Comment:
Would have been nice to batch these requests instead of 4 separate requests

Reply:
refactored with multicall